In [2]:
!pip install sklearn

  Running setup.py bdist_wheel for sklearn ... - done
  Stored in directory: /root/.cache/pip/wheels/76/03/bb/589d421d27431bcd2c6da284d5f2286c8e3b2ea3cf1594c074
Successfully built sklearn


In [0]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import f1_score, classification_report, accuracy_score, mean_squared_error, confusion_matrix

In [0]:
# reading in Dataset
paths = 'http://bigdata.cs.byu.edu/Lectures/EEGEyeState.csv'
df = pd.read_table(paths,sep=',',encoding='latin1')

# split dependent and independent variables
y = df['eyeDetection']
X = df.iloc[:,:-1]

# create test and train.
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=.3,random_state=40)

In [5]:
# XGBoost

from xgboost import XGBClassifier

xgb = XGBClassifier()
xgb.fit(X_train, np.ravel(y_train))

print(xgb.score(X_test, y_test))

print(sorted(zip(xgb.feature_importances_, X_test.columns), reverse=True))

print(confusion_matrix(y_test, xgb.predict(X_test)))

0.7975077881619937
[(0.1518405, 'P7'), (0.13957055, 'O1'), (0.1303681, 'F7'), (0.108895704, 'F8'), (0.0690184, 'AF4'), (0.05981595, 'T7'), (0.058282208, 'AF3'), (0.056748465, 'O2'), (0.055214725, 'FC6'), (0.04447853, 'FC5'), (0.041411042, 'F4'), (0.036809817, 'T8'), (0.030674847, 'F3'), (0.016871165, 'P8')]
[[2144  332]
 [ 578 1440]]


/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


In [13]:
param = {'n_estimators': [100, 150, 250],
        'max_depth': [10, 30, 50]}

gs = GridSearchCV(xgb, param, cv=5, n_jobs=-1)
gs_fit = gs.fit(X_train, np.ravel(y_train))
pd.DataFrame(gs_fit.cv_results_).sort_values('mean_test_score', ascending=False)

/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this wi

mean_fit_time  mean_score_time  mean_test_score  mean_train_score  \
2       5.817565         0.082746         0.937345          1.000000   
8       9.530825         0.111834         0.937154          1.000000   
5       9.551547         0.113545         0.936105          1.000000   
7       6.570726         0.074375         0.933340          1.000000   
4       6.512904         0.079634         0.932672          1.000000   
1       3.637878         0.048388         0.930860          0.999952   
3       4.755008         0.058424         0.929048          1.000000   
6       4.734123         0.054966         0.929048          1.000000   
0       2.511647         0.043527         0.924948          0.998093   

  param_max_depth param_n_estimators                                  params  \
2              10                250  {'max_depth': 10, 'n_estimators': 250}   
8              50                250  {'max_depth': 50, 'n_estimators': 250}   
5              30                250  {'max_depth': 30, 'n_estimators': 250}   
7              50                150  {'max_depth': 50, 'n_estimators': 150}   
4              30                150  {'max_depth': 30, 'n_estimators': 150}   
1              10                150  {'max_depth': 10, 'n_estimators': 150}   
3              30                100  {'max_depth': 30, 'n_estimators': 100}   
6              50                100  {'max_depth': 50, 'n_estimators': 100}   
0              10                100  {'max_depth': 10, 'n_estimators': 100}   

   rank_test_score  split0_test_score  split0_train_score       ...         \
2                1           0.945186            1.000000       ...          
8                2           0.949952            1.000000       ...          
5                3           0.946616            1.000000       ...          
7                4           0.947569            1.000000       ...          
4                5           0.946616            1.000000       ...          
1                6           0.935653            0.999762       ...          
3                7           0.944233            1.000000       ...          
6                7           0.944233            1.000000       ...          
0                9           0.933270            0.997377       ...          

   split2_test_score  split2_train_score  split3_test_score  \
2           0.934192            1.000000           0.930854   
8           0.933238            1.000000           0.927515   
5           0.933238            1.000000           0.927515   
7           0.923224            1.000000           0.924177   
4           0.923224            1.000000           0.924177   
1           0.929423            1.000000           0.923224   
3           0.920839            1.000000           0.918932   
6           0.920839            1.000000           0.918932   
0           0.924177            0.998331           0.916071   

   split3_train_score  split4_test_score  split4_train_score  std_fit_time  \
2            1.000000           0.937053             1.00000      0.067981   
8            1.000000           0.929423             1.00000      0.664029   
5            1.000000           0.929423             1.00000      0.057854   
7            1.000000           0.928469             1.00000      0.034289   
4            1.000000           0.928469             1.00000      0.044336   
1            1.000000           0.929900             1.00000      0.015519   
3            1.000000           0.921793             1.00000      0.051052   
6            1.000000           0.921793             1.00000      0.026705   
0            0.999285           0.921793             0.99845      0.016659   

   std_score_time  std_test_score  std_train_score  
2        0.009963        0.004859         0.000000  
8        0.011360        0.008987         0.000000  
5        0.003791        0.007685         0.000000  
7        0.001782        0.010107         0.000000  
4        0.004615      

In [6]:
# XGBoost after tuning

xgb = XGBClassifier(n_estimators=250, max_depth=15,
                             random_state=7)
xgb.fit(X_train, np.ravel(y_train))

print("Test score:")
print(xgb.score(X_test, y_test))

print("Importances:")
print(sorted(zip(xgb.feature_importances_, X_test.columns), reverse=True))

print("Confusion Matrix:")
cmat = confusion_matrix(y_test, xgb.predict(X_test))
mats = pd.DataFrame(cmat)
print(mats)
  
print("Sensitivity")
print(mats.iloc[1,1]/( mats.iloc[1,1]+ mats.iloc[0,1]))
print("Specificity")
print(mats.iloc[0,0]/( mats.iloc[0,0]+ mats.iloc[0,1]))

Train score:


/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


1.0
Test score:
0.9417000445037829
Importances:
[(0.09724568, 'O1'), (0.092441626, 'F7'), (0.09022885, 'F8'), (0.08030047, 'FC6'), (0.07473942, 'AF4'), (0.07235195, 'O2'), (0.06871251, 'T7'), (0.067868166, 'AF3'), (0.06518954, 'T8'), (0.06460723, 'FC5'), (0.061637454, 'F3'), (0.059453793, 'F4'), (0.05910441, 'P7'), (0.046118908, 'P8')]
Confusion Matrix:
      0     1
0  2362   114
1   148  1870
Sensitivity
0.9425403225806451
Specificity
0.9539579967689822


/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


# Summary

After splitting the model into train and test sets, we used scikit-learn's grid search functionality to test our XGBoost model with several different hyperparameter configurations. The best results were found using 250 trees with a max depth of 15.

With these settings, we were able to acieve 94% accuracy while maintaining fairly good precision and recall.


